### PR TITLE
Issue #481: Add support for probability aggregation to `expected_obs()`

### LINKED ISSUE #481: Add support for aggregation of probabilities in `expected_obs`
Toward using delay probability aggregation to encode a fixed reporting cycle into `epinowcast`, the Stan function `expected_obs` will need to be modified to take in an argument specifying which dates to be aggregated and which dates they should be aggregated to. This will likely be a square matrix with dimension `max_delay` which, when right-multiplied to the delay probability vector, will output new delay probabilities aggregated to the appropriate reporting days.

### PR DESCRIPTION
**Pull request by jessalynnsebastian**
## Description

This PR closes #481. It adds an indicator (`agg_probs`) and a matrix argument (`agg_indicator`) to `expected_obs()` that indicate whether to aggregate and how the aggregation should be performed, respectively. The matrix argument is structured to be directly multiplied to the probability vector: the rows represent which index the aggregated probability should be placed on in the new probability vector that accounts for the reporting cycle, and the columns represent whether a given probability in the probability vector should be aggregated. For example, if we had the aggregation matrix 

```math
\texttt{agg}\_\texttt{indicator} = \begin{bmatrix} 0 & 0 & 0 & 0 & 0 & 0 \\\ 0 & 0 & 0 & 0 & 0 & 0 \\\ 1 & 1 & 1 & 0 & 0 & 0 \\\ 0 & 0 & 0 & 0 & 0 & 0 \\\ 0 & 0 & 0 & 0 & 0 & 0 \\\ 0 & 0 & 0 & 1 & 1 & 1 \end{bmatrix}
```
the first 3 elements of the probability vector would be summed and placed on the third element of the new probability vector, and the last three elements would be summed and placed on the sixth element.

For now, `agg_probs` is set to 0 (for no aggregation) and `agg_indicator` is set to [0] in `expected_obs_from_index()`, since the infrastructure to implement this method doesn't yet exist.

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [x] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] My code follows the established coding standards.
- [x] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

### CODE REVIEW

File: inst/stan/functions/expected_obs.stan [hunk 1]
```diff
@@ -58,37 +62,38 @@
...
  * rep_lh[c(2, 20)] = -2
- * round(exp(expected_obs(tar_obs, equal_lh + rep_lh, 0)), 3)
+ * round(exp(expected_obs(tar_obs, equal_lh + rep_lh, 0, 0, [0])), 3)
  * # 0.033 0.005 0.034 0.034 0.034 0.202 0.027 0.027 0.027 0.027 0.027 0.151
  * # 0.021 0.021 0.021 0.106 0.014 0.014 0.014 0.002 0.016 0.016 0.016 0.016
  * # 0.016 0.016 0.016 0.016 0.016 0.016
  *
  * Example combining both date of reference and date of report effects
- * eobs <- exp(expected_obs(tar_obs, ref_lh + rep_lh, 0))
+ * eobs <- exp(expected_obs(tar_obs, ref_lh + rep_lh, 0, 0, [0]))
```
Comment (seabbs) [inline inst/stan/functions/expected_obs.stan:71]: Can you add an example here showing what this does (i.e below this example) when it is on

**Review by seabbs — CHANGES_REQUESTED** [review]
Nice this looks good to me. A few minor comments for expanding the documentation. For this to be ready to merge I think we just need a news item linked to it

File: tests/testthat/test-stan_expected_obs.R [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+enw_stan_to_r(
+  c("expected_obs.stan", "hazard.stan"),
+  system.file("stan", "functions", package = "epinowcast")
+)
+# note these tests require enw_stan_to_r() to be run first
+
+test_that("expected_obs() aggregates probabilities correctly", {
+  tar_obs <- 0
+  lh <- c(0.1, 0.2, 0.3, 0.4)
```
Comment (jessalynnsebastian) [inline tests/testthat/test-stan_expected_obs.R:14]: Will also need to edit the test case in light of the log-prob vs. prob mistake

File: inst/stan/functions/expected_obs.stan [hunk 3]
```diff
@@ -58,37 +65,38 @@
...
     p = inv_logit(lh);
     }
     profile("model_likelihood_expected_obs_hazard_to_prob") {
     p = hazard_to_log_prob(p);
     }
-    profile("model_likelihood_expected_obs_prod_p") {
+  }
+  if (agg_probs == 1) {
+    p = agg_indicator * p;
+  }
```
[the remark this answers is not in the document]
Comment (seabbs) [reply inst/stan/functions/expected_obs.stan]: Great catch

File: tests/testthat/test-stan_expected_obs.R (hunk 2 above)
Comment (seabbs) [reply tests/testthat/test-stan_expected_obs.R]: Worth thinking about what dev steps could have taken to spot this sooner (though given this is still in draft its pretty good). For me I didn't run the test case - just sight read which I think would have helped me catch it

File: inst/stan/functions/expected_obs.stan [hunk 4]
```diff
@@ -12,8 +12,15 @@
  * a given day. When `ref_as_p` is 0, this should be the logit hazard instead
  * of probability.
  * 
- * @param ref_as_p An integer flag (0 or 1) indicating whether the reference date input should be treated as a logit hazard or probability. Set to 1 when
+ * @param ref_as_p An integer flag (0 or 1) indicating whether the reference dateinput should be treated as a logit hazard or probability. Set to 1 when
```
Comment (seabbs) [inline inst/stan/functions/expected_obs.stan:15]:
```suggestion
 * @param ref_as_p An integer flag (0 or 1) indicating whether the reference date input should be treated as a logit hazard or probability. Set to 1 when
```

File: inst/stan/functions/expected_obs.stan [hunk 5]
```diff
@@ -58,37 +65,51 @@
...
  * # 0.007 0.007 0.006 0.024 0.003 0.003 0.003 0.000 0.003 0.002 0.002 0.002
  * # 0.002 0.002 0.002 0.002 0.002 0.002
+ *
+ * Example with aggregation of probabilities where aggregation occurs on every fifth day
+ * agg_probs <- matrix(c(rep(0, times = 30 * 4),
+ *                       rep(c(rep(1, times = 5),
+ *                             rep(0, times = 6 * 5 + 30 * 4)),
+ *                           times = 5),
+ *                       rep(1, times = 5)), ncol = 30, byrow = TRUE)
+ * eobs <- exp(expected_obs(tar_obs, date_p + rep(0, 30), 1, 1, agg_probs))
```
Comment (seabbs) [inline inst/stan/functions/expected_obs.stan:87]: when I ran this locally I found having `exp_obs |> exp() |> plot()` this to be helpful. This looks like it produces the expected output and that output makes sense to me.

File: inst/stan/functions/expected_obs.stan [hunk 6]
```diff
@@ -58,37 +65,51 @@
...
     profile("model_likelihood_expected_obs_inv_logit") {
     p = inv_logit(lh);
     }
     profile("model_likelihood_expected_obs_hazard_to_prob") {
     p = hazard_to_log_prob(p);
     }
-    profile("model_likelihood_expected_obs_prod_p") {
+  }
+  if (agg_probs == 1) {
+    p = exp(p);
```
Comment (seabbs) [inline inst/stan/functions/expected_obs.stan:107]: I didn't gain a lot from having this over 3 lines. What do you think about making it all one line? 

I think we should have an issue for exploring other approaches that are potentially more performant (but not doing now). In my head those would be: 

1. The log_sum_exp approach looping over the rows
2. A sparse matrix approach with the current exp -> matrix -> log

**Review by seabbs — DISMISSED** [review]
This looks great. Good catch on the issue. I have a few style comments but otherwise I think this is good to go.

File: inst/stan/functions/expected_obs.stan (hunk 6 above)
Comment (jessalynnsebastian) [reply inst/stan/functions/expected_obs.stan:107]: Yeah this is fair enough, I'll condense. Will also write an issue for your suggestions here

**Review by seabbs — APPROVED** [review]
Nice this looks good to me!

**Comment by seabbs** [comment]
(I pinged myself for review as it looked like you addressed all the comments. For the future to get peoples attention it can be helpful if the PR author does this when they are ready :) )

**Pull request merged**